Fix caret not auto scroll to visible when page has multiple editors

theachoem opened the pull request:
<!-- 
Briefly describe your changes and summarize in the title.
Contributor Guide: https://github.com/singerdmx/flutter-quill/blob/master/CONTRIBUTING.md
Package versioning is automated.
Add updates to `Unreleased` in `CHANGELOG.md` following [Keep a Changelog](https://keepachangelog.com/en/1.1.0/) format.
-->

## Description

For this fix, I copied all code based on flutter `editable_text.dart` to have the same behavior. I have pushed multiple commits to track the changes:
- The first commit is just copied original code from Flutter (no change): [editable_text.dart#L4228](https://github.com/flutter/flutter/blob/3.29.0/packages/flutter/lib/src/widgets/editable_text.dart#L4228)
- The second commit is refactoring to fit flutter_quill


## Related Issues
- https://github.com/singerdmx/flutter-quill/issues/2569

## Fixed Demo

https://github.com/user-attachments/assets/ae027618-21c6-499f-a3e8-4bc338f79a18

## Type of Change

<!---
Check the boxes that apply with x and leave the others empty. For example:
- [ ] ✨ **New feature:** Adds new functionality without breaking existing features.
- [x] 🛠️ **Bug fix:** Resolves an issue without changing current behavior.
-->

- [ ] ✨ **Feature:** New functionality without breaking existing features.
- [x] 🛠️ **Bug fix:** Resolves an issue without altering current behavior.
- [ ] 🧹 **Refactor:** Code reorganization, no behavior change.
- [ ] ❌ **Breaking:** Alters existing functionality and requires updates.
- [ ] 🧪 **Tests:** New or modified tests
- [ ] 📝 **Documentation:** Updates or additions to documentation.
- [ ] 🗑️ **Chore:** Routine tasks, or maintenance.
- [ ] ✅ **Build configuration change:** Build/configuration changes.

theachoem committed:
Copy _scheduleShowCaretOnScreen from flutter editable_text

theachoem committed:
Refactor _scheduleShowCaretOnScreen to fit with flutter_quill

theachoem commented:
Mostly work, but ran into a bit of issue, it does not auto scroll until I start typing:

https://github.com/user-attachments/assets/0e0062cd-2367-4757-a626-f4b1f47b6b51

theachoem commented:
Mostly work, but ran into a bit of issue, it does not auto scroll until I start typing:


https://github.com/user-attachments/assets/63dd56bc-69f4-4443-8928-dc4eb1687293

theachoem committed:
Handle didChangeMetrics (handle show caret during rotate & initial focus)

theachoem commented:
> Mostly work, but ran into a bit of issue, it does not auto scroll until I start typing:
> 
>  Screen.Recording.2025-05-18.at.1.54.38.in.the.morning.mov

Trigger show caret method in didChangeMetrics fixed the issue.


https://github.com/user-attachments/assets/9f3f7923-bc53-4a76-9360-636415a1ea83

theachoem commented:
> Mostly work, but ran into a bit of issue, it does not auto scroll until I start typing:


https://github.com/user-attachments/assets/1bfd2b22-85a2-4bfc-8e8a-c72bce463440

theachoem commented:
> Mostly works, but ran into a bit of an issue, it does not auto-scroll on the first request focus until I start typing:
> https://github.com/user-attachments/assets/63dd56bc-69f4-4443-8928-dc4eb1687293

Handle didChangeMetrics fix the issue:

https://github.com/user-attachments/assets/1bfd2b22-85a2-4bfc-8e8a-c72bce463440

Reference:
- [editable_text.dart#L4307](https://github.com/flutter/flutter/blob/3.29.0/packages/flutter/lib/src/widgets/editable_text.dart#L4307)

theachoem commented:
@EchoEllet I have released this to my production app for a few days now. If you have any concerns about it being merged, let me know. I will refactor accordingly.

EchoEllet commented on CHANGELOG.md:
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
+- Fix caret not auto scroll to visible when page has multiple editors [#2570](https://github.com/singerdmx/flutter-quill/pull/2570).
+  **BREAKING**: QuillEditorConfig is now accept `EdgeInsets padding` instead of `EdgeInsetsGeometry`.
```
Is this change related to this fix? If not, can we move it into a separate PR?

theachoem replied:
It is required because we need `padding.bottom` which is only available for EdgeInsets. https://github.com/singerdmx/flutter-quill/pull/2570/files#diff-5f87ac57652e1ab8fa6a4062411e9cd842ab7a8163fe0b74a1f82fa7eb8f494cR1249.

This is entirely base on flutter.

EchoEllet replied:
> It is required because we need padding.bottom which is only available for EdgeInsets.

Is this required for the fix itself?

theachoem replied:
Yes!

EchoEllet commented on CHANGELOG.md:
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+- Fix caret not auto scroll to visible when page has multiple editors [#2570](https://github.com/singerdmx/flutter-quill/pull/2570).
+  **BREAKING**: QuillEditorConfig is now accept `EdgeInsets padding` instead of `EdgeInsetsGeometry`.
```
The first change should be under `Fixed` and the second under the `Changed` section.

EchoEllet reviewed (commented):
There are a lot of changes that I'm not related to at the moment, I'm not able to work on the project for a while (again).

Unfortunately, I can't take responsibility for reviewing PRs at this point.

Hopefully, fellow maintainers will review it at some point.

theachoem commented:
@EchoEllet I understand. I am confident about it, and I think if other contributors do it, it will likely output the same result. Let  me know what I can do.

theachoem committed:
Fix changelog

theachoem committed:
Fix doc import cannot have combinators

warning • Doc imports can't have show or hide combinators • lib/src/editor/config/editor_config.dart:1:42 • doc_import_cannot_have_combinators
warning • Doc imports can't have show or hide combinators • lib/src/toolbar/config/buttons/link_style_options.dart:1:45 • doc_import_cannot_have_combinators

Solution is to remove 'show', just @docImport '../path' is enough to import and make type clickable in comment. This issue raise latest flutter 3.32.0

singerdmx commented:
@EchoEllet shall we merge this?

EchoEllet commented:
I'm not able to review it in a reasonable time frame, sorry but I can't confirm my review so you need to make the call.

theachoem commented:
@anneinge95 works fine for me, make sure you put the toolbar in the bottom navigation.

EchoEllet commented on CHANGELOG.md:
```diff
@@ -10,6 +10,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+- Fix caret not auto scroll to visible when page has multiple editors [#2570](https://github.com/singerdmx/flutter-quill/pull/2570).
+
+### Changed
+- **BREAKING**: QuillEditorConfig is now accept `EdgeInsets padding` instead of `EdgeInsetsGeometry`.
```
Why was this breaking change introduced? Does not seem to be related to the fix, and the PR description does not state that this is a breaking change.

theachoem commented:
@EchoEllet do I need to redo it? padding breaking is absolutely necessary, I already replied the reason in the review 💔

EchoEllet commented:
> let me know what I can help.

This isn't an issue with your PR; it's a project maintenance issue.

>  However, all the changes in that PR are necessary and part of the intended fix

The lint fixes are not part of the PR fix, and it's a CI issue.

> For this fix, I copied all code based on flutter editable_text.dart to have the same behavior.

Generally, I prefer to avoid the approach of copying complex code and modifying it. It was done previously with the magnifier feature, but [had to revert it](https://github.com/singerdmx/flutter-quill/pull/2413) due to significant regressions.

I'm no longer a project maintainer, but
It looks like PRs are being merged without any review, consideration, testing, or documentation, so every consumer is a testing site. I have seen this pattern before when I was a maintainer, and I still see it today.

Unfortunately, it looks like there are no longer active project maintainers, so for new projects I highly recommend alternatives such as [`super_editor`](https://pub.dev/packages/super_editor) or [`appflowy_editor`](https://pub.dev/packages/appflowy_editor).

EchoEllet commented:
Merging it might work for your app, but without proper testing, I can't risk merging it for everyone. I'm not the original maintainer and trying to leave—please consider forking if you need it.

>  But Flutter Quill already copies a lot from Flutter’s text field

We should improve the situation instead of blindly copy-pasting code we don't understand (Flutter internal) and shipping it to the users of this package.

> I agree, but if I don’t fix the lint, the CI fails and the PR can’t be merged.

This can be fixed without introducing the change as part of this PR. This makes the revert (if we decided to make it) harder with many conflicts.

> That might be true, but the project still works well for my app, 

You could use a dependency override with your fork, and pull changes from upstream repo to upgrade.

<details><summary>To explain the situation (off-topic)</summary>
<p>

I wanted to quit Flutter Quill 1 or 2 years ago, but all the previous maintainers have abandoned the project, so I'm the only one who is working with the project (see commit history).

I don't think we're as strict as we used to be. PRs are being merged without enough consideration, and when regressions happen, I'm often the only one left to fix them and clean up the mess. Maintaining a codebase like Flutter Quill is time-consuming for anyone who has worked with its internals.

</p>
</details>
